### PR TITLE
Subscription refactoring

### PR DESCRIPTION
**Pull request by mkatliar**
Main improvements:
 * RAII-based subscription management
 * Exception-based error handling in many places

### CODE REVIEW

**Review by Wojcik98 — CHANGES_REQUESTED** [review]
Next time please make several smaller PRs :)

File: src/rws_interface.cpp [hunk 1]
```diff
@@ -569,274 +558,247 @@ std::vector<RobotWareOptionInfo> RWSInterface::getPresentRobotWareOptions()
...
+      !dynamic_info.status.empty() &&
+      !dynamic_info.jog_mode.empty() &&
+      dynamic_info.mode != MechanicalUnitMode::UNKNOWN_MODE)
+  {
+    ;
+  }
+  else
+  {
+    throw std::logic_error("RWSInterface::getMechanicalUnitDynamicInfo: inconsistent data");
+  }
```
[the remark this answers is not in the document]
Comment (mkatliar) [reply src/rws_interface.cpp:680]: done

File: include/abb_librws/rws_client.h [hunk 2]
```diff
@@ -608,19 +373,19 @@ class RWSClient : public POCOClient
    * \param resource specifying the file's directory and name.
    * \param replace indicating if the actual module into the controller must be replaced by the new one or not.
    *
-   * \return RWSResult containing the result.
+   * \throw std::exception if something goes wrong.
    */
-  RWSResult loadModuleIntoTask(const std::string& task, const FileResource& resource, const bool replace = false);
+  void loadModuleIntoTask(const std::string& task, const FileResource& resource, const bool replace = false);
```
[the remark this answers is not in the document]
Comment (mkatliar) [reply include/abb_librws/rws_client.h:378]: done

File: src/rws_common.cpp [hunk 3]
```diff
@@ -201,6 +201,37 @@ bool xmlNodeHasAttribute(const Poco::XML::Node* p_node, const std::string& name,
...
+  std::string quot = "&quot;";
+  size_t quot_position = 0;
+  do
+  {
+    quot_position = result.find(quot);
+    if (quot_position != std::string::npos)
+    {
+      result.replace(quot_position, quot.size(), "");
+    }
+  } while (quot_position != std::string::npos);
```
[the remark this answers is not in the document]
Comment (mkatliar) [reply src/rws_common.cpp:231]: @Wojcik98 sorry, I could not understand your comment. Could you rephrase it?
This function was just copied "as is" to here from some other part of the code, because it seems to belong here.

File: src/rws_subscription.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  {
+  }
+  
+  
+  bool SubscriptionReceiver::waitForEvent(SubscriptionEvent& event)
+  {
+    WebSocketFrame frame;
+    if (webSocketReceiveFrame(frame))
+    {
+      // std::clog << "WebSocket frame received: flags=" << frame.flags << ", frame_content=" << frame.frame_content << std::endl;
```
[the remark this answers is not in the document]
Comment (mkatliar) [reply src/rws_subscription.cpp:132]: done

File: src/rws_subscription.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+      Poco::AutoPtr<Poco::XML::Document> doc = parser_.parseString(frame.frame_content);
+
+      event.value = xmlFindTextContent(doc, XMLAttribute {"class", "lvalue"});
+
+      // IMPORTANT: don't use AutoPtr<XML::Node> here! Otherwise you will get memory corruption.
+      if (Poco::XML::Node const * node = doc->getNodeByPath("html/body/div/ul/li/a"))
```
[the remark this answers is not in the document]
Comment (mkatliar) [reply src/rws_subscription.cpp:138]: This is a valid syntax -- it makes the `node` variable visible only inside the `if` block.

File: include/abb_librws/rws_state_machine_interface.h [hunk 6]
```diff
@@ -1124,387 +1094,289 @@ class RWSStateMachineInterface : public RWSInterface
...
-       * \return bool indicating if the communication was successful or not.
        */
-      bool rightVacuum2On() const;
+      void rightVacuum2On() const;
 
       /**
        * \brief Signal the StateMachine to run SmartGripper routine(s).
        *
-       * \return bool indicating if the signaling was successful or not.
+       * \return void indicating if the signaling was successful or not.
```
[the remark this answers is not in the document]
Comment (mkatliar) [reply include/abb_librws/rws_state_machine_interface.h:1344]: :-)
done

File: src/rws_common.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Wojcik98) [reply src/rws_common.cpp:231]: The function is called `findSubstringContent`, which suggest that it will just return the content found between `substring_start` and `substring_end`. However, this function first finds this content, but then removes from it all substrings `&quot`.
My suggestion was to split it into two functions: first one to just find the content and the other to just remove `&quot` from input string.
But if it was copied over then I guess we can leave it as it is.

File: src/rws_subscription.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Wojcik98) [reply src/rws_subscription.cpp:137]: Ok then, I wasn't familiar with it and suspected this syntax isn't "clean" :)
On the second thought though it's kinda similar to variable initialization in `for` loop, nice feature.

**Review by Wojcik98 — APPROVED** [review]
